chain/events/filter/index.go

```diff
@@ -530,16 +705,45 @@ func (ei *EventIndex) CollectEvents(ctx context.Context, te *TipSetEvents, rever
 
 	// lets handle the revert case first, since its simpler and we can simply mark all events in this tipset as reverted and return
 	if revert {
+		tsKeyCid, err := te.msgTs.Key().Cid()
+		if err != nil {
+			return xerrors.Errorf("tipset key cid: %w", err)
+		}
+
 		_, err = tx.Stmt(ei.stmtRevertEventsInTipset).Exec(te.msgTs.Height(), te.msgTs.Key().Bytes())
 		if err != nil {
 			return xerrors.Errorf("revert event: %w", err)
 		}
 
+		_, err = tx.Stmt(ei.stmtRevertEventSeen).Exec(te.msgTs.Height(), tsKeyCid.Bytes())
+		if err != nil {
+			return xerrors.Errorf("revert event seen: %w", err)
+		}
+
 		err = tx.Commit()
 		if err != nil {
 			return xerrors.Errorf("commit transaction: %w", err)
 		}
 
+		ei.mu.Lock()
+		tSubs := ei.updateSubs
```

Comment:
this isn't a full copy, you'll need to go further than this to be protected

```suggestion
		tSubs := make([]int, len(ei.updateSubs))
		copy(tSubs, ei.updateSubs)
```

Reply:
could make this a private method `updateSubsCopy()` or similar, since it's done twice in here

Reply:
Done.